src/main/java/com/googleinterns/zoomtube/servlets/TranscriptServlet.java

```diff
@@ -37,42 +53,133 @@
+  public static final String ATTR_DURATION = "dur";
+  public static final String TAG_TEXT = "text";
+  public static final String PARAM_LECTURE = "lecture";
+  public static final String PARAM_LECTURE_ID = "id";
+  public static final String PARAM_VIDEO_ID = "video";
+
+  private DatastoreService datastore;
 
   @Override
   public void init() throws ServletException {
-    // TODO: Implement Transcript.
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  /**
+   * Initializes the servlet with {@code testDatastore} created during testing.
+   *
+   * <p>The unit tests need access to the datastore to check that doPost() puts
+   * the entities in datastore.
+   */
+  @VisibleForTesting
+  void init(DatastoreService testDatastore) {
+    datastore = testDatastore;
   }
 
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // TODO: Pass the video ID from another servlet.
-    String transcriptXMLUrl = TRANSCRIPT_XML_URL_TEMPLATE + TEST_VIDEO_ID;
-    DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
-    final DocumentBuilder documentBuilder;
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Document document = getTranscriptXmlAsDocument(request).get();
+    putTranscriptLinesInDatastore(request, document);
+  }
+
+  /**
+   * Returns the transcript for a video as a document. Otherwise, returns Optional.empty()
+   * if there is a parsing error.
+   *
+   * @param request Indicates the video to extract the transcript from.
+   */
+  private Optional<Document> getTranscriptXmlAsDocument(HttpServletRequest request)
+      throws IOException {
+    String videoId = request.getParameter(PARAM_VIDEO_ID);
+    String transcriptXMLUrl = TRANSCRIPT_XML_URL_TEMPLATE + videoId;
+
     final Document document;
     try {
-      documentBuilder = documentBuilderFactory.newDocumentBuilder();
+      DocumentBuilder documentBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
       document = documentBuilder.parse(new URL(transcriptXMLUrl).openStream());
       document.getDocumentElement().normalize();
     } catch (ParserConfigurationException | SAXException e) {
       // TODO: Alert the user.
       System.out.println("XML parsing error");
-      return;
+      return Optional.empty();
     }
-    NodeList nodeList = document.getElementsByTagName(TEXT_TAG);
+    return Optional.of(document);
+  }
+
+  /**
+   * Puts each transcript line from {@code document} in datastore as its own entity.
+   *
+   * @param request Indicates the lecture key to group the transcript lines under.
+   * @param document The XML file containing the transcript lines.
+   */
+  private void putTranscriptLinesInDatastore(HttpServletRequest request, Document document) {
+    long lectureId = Long.parseLong(request.getParameter(PARAM_LECTURE_ID));
```

Comment:
instead of passing the request, pass the lectureId. This method doesn't care what the request is. It only cares about lectureId.

Reply:
done!